### PR TITLE
#22425 NavTool method getNav must return only published pages

### LINKED ISSUE #22425: NavTool method getNav must return only published pages.
**Describe the bug**
When $navtool.getNav("$VTLSERVLET_URI") method is called. It returns unpublished pages as well.
Related Ticket: https://dotcms.zendesk.com/agent/tickets/107632
Reproduced on: 22.05

**To Reproduce**
1. log in to [Demo Site](https://demo.dotcms.com/dotAdmin/#/public/login?r=1655383397360).
2. Go to Site->Browser portlet.
3. Go to application/vtl/carousel directory.
4. open banner-carousel.vtl.
5. Add the following code.
```
#set($contextualNav = $navtool.getNav("$VTLSERVLET_URI"))

#if($contextualNav != false)
#foreach($item in $contextualNav)
<li class="level_1#if($item.active) current#end #if($UtilMethods.isSet($item.children)) parent#end"><a
        href="${item.href}" data-gtmconversions="subpage">$item.title</a></li>
#end
#end
```
6. Go to the destinations directory.
7. Create a new page name 'test' in the destinations directory, and mark the 'show on menu' option.
8. Click on the save button (not publish).
9. Now go to [this page](https://demo.dotcms.com/destinations/).
10. Verify the result unpublished page appears.

**Expected behavior**
When $navtool.getNav is called only published pages must appear in the result set.

**Additional context**
The possible source of the issue is [#21385](https://github.com/dotCMS/core/issues/21385)

### PR DESCRIPTION
**Pull request by fabrizzio-dotCMS**
### Proposed Changes
* getNav only grabs stuff from the cache when visiting a Live page
* We exclude unpublished stuff only when viewing a Live page.
* getNav is built on the fly for EDIT_MODE and PREVIEW_MODE to stay consistent with the page editor

### CODE REVIEW

File: dotCMS/src/main/java/com/dotcms/rest/api/v1/page/NavResource.java [hunk 1]
```diff
@@ -116,7 +118,8 @@ public final Response loadJson(@Context final HttpServletRequest request, @Conte
                     response, Config.CONTEXT);
 
             final String path = (!uri.startsWith("/")) ? "/" + uri : uri;
-
+            //Force NavTool to behave as Live when rendering items
+            request.setAttribute(WebKeys.PAGE_MODE_PARAMETER, PageMode.LIVE);
```
Comment (freddyucv) [inline dotCMS/src/main/java/com/dotcms/rest/api/v1/page/NavResource.java:122]: why do we need save the PageMode as a request attribute? we have it here

https://github.com/dotCMS/core/blob/d3a5a862e08c6111fa22b21b7ceba8e5e21e8a3b/dotCMS/src/main/java/com/dotmarketing/util/PageMode.java#L58

**Review by freddyucv — CHANGES_REQUESTED** [review]

**Review by nollymar — APPROVED** [review]

File: dotCMS/src/main/java/com/dotcms/rest/api/v1/page/NavResource.java (hunk 1 above)
Comment (fabrizzio-dotCMS) [reply dotCMS/src/main/java/com/dotcms/rest/api/v1/page/NavResource.java:122]: now NavTool internally uses PageMode. And it works correctly when invoked from a page 
When invoked from a rest endpoint. it needs to resolve as LIVE. But it isn't therefore  I needed to override the attribute on the request so it resolves to LIVE. @freddyucv that's why

File: dotCMS/src/main/java/com/dotcms/rendering/velocity/viewtools/navigation/NavTool.java [hunk 2]
```diff
@@ -73,153 +70,185 @@ protected NavResult getNav(Host host, String path) throws DotDataException, DotS
...
+                    .map(o -> (Versionable) o)
+                    .filter(versionable -> Try.of(() -> !versionable.isLive()).getOrElse(false))
+                    .collect(Collectors.toList());
+            menuItems.removeAll(nonLive);
         }
+        return menuItems;
+    }
+
+    PageMode pageMode(){
+        final HttpServletRequest request = HttpServletRequestThreadLocal.INSTANCE.getRequest();
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/rendering/velocity/viewtools/navigation/NavTool.java:250]: why do you need this thing? a View tool always has the request available.

File: dotCMS/src/main/java/com/dotcms/rest/api/v1/page/NavResource.java (hunk 1 above)
Comment (fabrizzio-dotCMS) [reply dotCMS/src/main/java/com/dotcms/rest/api/v1/page/NavResource.java]: done @freddyucv

File: dotCMS/src/main/java/com/dotcms/rendering/velocity/viewtools/navigation/NavTool.java (hunk 2 above)
Comment (fabrizzio-dotCMS) [reply dotCMS/src/main/java/com/dotcms/rendering/velocity/viewtools/navigation/NavTool.java:250]: it's a convenient method. nothing wrong there IMO

File: dotCMS/src/main/java/com/dotcms/rendering/velocity/viewtools/navigation/NavTool.java [hunk 3]
```diff
@@ -73,153 +70,185 @@ protected NavResult getNav(Host host, String path) throws DotDataException, DotS
...
+        //It makes no sense putting things in cache when we're browsing in Preview or Edit mode
+        if(pageMode.showLive) {
+            navToolCache.putNav(host.getIdentifier(), folder.getInode(), result, languageId);
+        }
+        return new NavResultHydrated(result, this.context);
+    }
 
-            return new NavResultHydrated(result, this.context);
+    private List<?> filterNonLiveItems(final List<?> menuItems) {
+        if (!menuItems.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (fabrizzio-dotCMS) [reply dotCMS/src/main/java/com/dotcms/rendering/velocity/viewtools/navigation/NavTool.java:238]: in my opinion, what you suggest is more invasive and would require a change in the folderAPI and potentially in the factory too. Here I can filter by the type of versionable I care about

**Review by freddyucv — APPROVED** [review]

**Review by jdotcms — APPROVED** [review]

**Pull request merged**